KAFKA-15917: Wait for zombie sink tasks' consumers to commit offsets before trying to modify their offsets in integration tests

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-15917)

In some rare cases, our integration tests observe that tasks are up and running for our sink connectors, and issue a subsequent request to stop the connector, before the consumers of these tasks have had a chance to form/join a group. When this happens, requests to alter/reset the offsets for the connector actually succeed, since the group either doesn't exist, or is completely empty.

As a quick fix, we can wait for offsets to be committed for the connector before stopping it and attempting to modify its offsets. This goes a bit further than necessary (we should only have to wait for at least one task's consumer to have formed/joined a group), but off the top of my head it was the cleanest and briefest way to guarantee that a group had been formed. It also makes a little more sense in the context of the test, since there's not much use in modifying connector offsets when none exist yet.

In addition, the assertion error messages generated when sink connectors do not catch up to the expected committed offsets in time are augmented with useful information about the expected and actual offsets.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
KAFKA-15524, KAFKA-15917: Wait for zombie sink tasks' consumers to commit offsets before trying to modify their offsets in integration tests

C0urante commented:
@yashmayya @gharris1727 If either of you have a moment, here's another flaky test fix. Would appreciate a look!

yashmayya commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java:
```diff
@@ -801,6 +806,11 @@ public void testResetSinkConnectorOffsetsZombieSinkTasks() throws Exception {
 
+        // Make sure the tasks' consumers have had a chance to actually form a group
+        // (otherwise, the reset request will succeed because there won't be any active consumers)
+        verifyExpectedSinkConnectorOffsets(connectorName, topic, 1, NUM_RECORDS_PER_PARTITION,
```
Looks like this test was already verifying this?

gharris1727 replied:
And since this commit no longer changes the behavior of the Reset test, does that mean that KAFKA-15224 is not addressed in this commit? Is that flaky failure still possible?

C0urante replied:
Great point, I've removed mention of KAFKA-15524 from this PR.

yashmayya reviewed (commented):
Thanks Chris, the fix here makes sense to me. Although the two linked Jira tickets seem to contain different failure scenarios to the one you describe in the PR description?

C0urante committed:
Remove redundant offset verification

C0urante committed:
Add useful error messages to verifyExpectedSinkConnectorOffsets

C0urante commented:
@gharris1727 sorry for the delay. You're correct that this PR did not actually address KAFKA-15224, and I'm honestly not sure how I missed that the first time around. I've updated the description to only mention KAFKA-15917, added some useful info to assertion messages when sink connectors fail to catch up to the expected committed offsets in these tests, and rebased on the latest trunk.

Would you mind giving this another pass when you have a moment? Currently this test is the flakiest of all Connect integration tests, failing nearly 10% of the time.

yashmayya commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/OffsetsApiIntegrationTest.java:
```diff
@@ -976,27 +988,53 @@ private String modifySinkConnectorOffsetsWithRetry(ConnectorOffsets offsetsToAlt
      * @param connectorName the name of the sink connector whose offsets are to be verified
      * @param expectedTopic the name of the Kafka topic that the sink connector is consuming from
-     * @param expectedPartitions the number of partitions that exist for the Kafka topic
+     * @param expectedPartitions the number of partitions that exist for the Kafka<<< topic
```
Stray change?

yashmayya reviewed (approved):
LGTM, thanks @C0urante!

C0urante committed:
Fix Javadoc

C0urante commented:
Thanks Yash!
